pkg/spoke/addon/registration_controller.go

```diff
@@ -0,0 +1,232 @@
+		SecretName:      config.secretName,
+		AdditonalSecretData: map[string][]byte{
+			clientcert.KubeconfigFile: c.kubeconfigData,
+		},
+	}
+
+	csrOption := clientcert.CSROption{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("addon-%s-%s-", c.clusterName, config.AddOnName),
+			Labels: map[string]string{
+				// the labels are only hints. Anyone could set/modify them.
+				clientcert.ClusterNameLabel: c.clusterName,
+				clientcert.AddonNameLabel:   config.AddOnName,
+			},
+		},
+		Subject:         config.x509Subject(c.clusterName, c.agentName),
+		SignerName:      config.SignerName,
+		EventFilterFunc: createCSREventFilterFunc(c.clusterName, config.AddOnName, config.SignerName),
+	}
+
+	clientCertController := clientcert.NewClientCertificateController(
+		clientCertOption,
+		csrOption,
+		c.hubCSRInformer,
+		c.hubCSRClient,
+		kubeInformerFactory.Core().V1().Secrets(),
+		c.kubeClient.CoreV1(),
+		c.recorder,
+		fmt.Sprintf("ClientCertController@addon:%s:signer:%s", config.AddOnName, config.SignerName),
+	)
+
+	go kubeInformerFactory.Start(ctx.Done())
+	go clientCertController.Run(ctx, 1)
+
+	c.stopRegistrationFuncs[config.hash] = stopFunc
+}
+
+// stopRegistration stops the client certificate controller for the given config
+func (c *addOnRegistrationController) stopRegistration(config registrationConfig, ctx context.Context) error {
+	if stopFunc, ok := c.stopRegistrationFuncs[config.hash]; ok {
+		// stop the client certificate controller
+		stopFunc()
+
+		// delete the secret generated
+		err := c.kubeClient.CoreV1().Secrets(config.InstallationNamespace).Delete(ctx, config.secretName, metav1.DeleteOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		delete(c.stopRegistrationFuncs, config.hash)
+	}
+	return nil
+}
+
+// cleanup cleans both the registration configs and client certificate controllers for the addon
+func (c *addOnRegistrationController) cleanup(addOnName string, ctx context.Context) error {
+	errs := []error{}
+	for _, config := range c.addOnRegistrationConfigs[addOnName] {
+		if err := c.stopRegistration(config, ctx); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if err := operatorhelpers.NewMultiLineAggregate(errs); err != nil {
+		return err
+	}
+
+	delete(c.addOnRegistrationConfigs, addOnName)
+	return nil
+}
+
+func createCSREventFilterFunc(clusterName, addOnName, signerName string) factory.EventFilterFunc {
+	return func(obj interface{}) bool {
+		accessor, err := meta.Accessor(obj)
+		if err != nil {
+			return false
+		}
+		labels := accessor.GetLabels()
+		// only enqueue csr from a specific managed cluster
+		if labels[clientcert.ClusterNameLabel] != clusterName {
+			return false
+		}
+		// only enqueue csr created for a specific addon
+		if labels[clientcert.AddonNameLabel] != addOnName {
+			return false
+		}
+
+		// only enqueue csr with a specific signer name
+		csr, ok := obj.(*certificatesv1beta1.CertificateSigningRequest)
```

Comment:
you should update to csr v1.  Is anything stopping you from doing that nowish?

Reply:
I'll do it in a separated task.